libraries/reconstruction_cuda/cuda_gpu_iirconvolve.cu

```diff
@@ -1,107 +1,301 @@
+using shared_type = T[BLOCK_SIZE][BLOCK_SIZE+1];
+template< typename T >
+using data_ptr = T * __restrict__;
+template< typename T >
+using warp_shared_type = volatile T[WARP_SIZE];
+
+__device__ float power(float base, int exp) {
+    return powf(base, exp);
+}
+
+__device__ double power(double base, int exp) {
+    return pow(base, exp);
+}
+
+template< typename T, typename IndexFunc >
+__device__ void load_warp(data_ptr< T > in, warp_shared_type< T > sdata, IndexFunc f) {
+    const T z = sqrt(3.0) - 2.0;
+    const T z_0 = (1.0 + z) / z;
+
+    int tid = threadIdx.x;
+
+    if (tid == 0) {
+        sdata[tid] = in[f(tid)] * z_0;
+    } else {
+        sdata[tid] = in[f(tid)] * power(z, tid);
     }
+}
+
+template< typename T >
+__device__ void sum_warp(data_ptr< T > in, warp_shared_type< T > sdata, int offset) {
+    const int tid = threadIdx.x;
+
+    if (tid < 16) sdata[tid] += sdata[tid + 16];
+    if (tid < 8)  sdata[tid] += sdata[tid + 8];
+    if (tid < 4)  sdata[tid] += sdata[tid + 4];
+    if (tid < 2)  sdata[tid] += sdata[tid + 2];
+
+    if (tid == 0) {
+        in[offset] = sdata[0] + sdata[1];
+    }
+}
+
+/*
+    For each row computes sum of the row weighted by some numbers,
+    because of limited precision of float/double, it is done only for
+    first 32 columns
+    Saves sum to first column in the row
+*/
+template< typename T >
+__global__ void sum_rows(data_ptr< T > in, int x) {
+    int row = blockIdx.x * x;
+    __shared__ warp_shared_type< T > sdata;
+
+    load_warp(in, sdata, [row](int tid) { return row + tid; });
+    sum_warp(in, sdata, row);
+}
 
+/*
+    Alternative of sum_rows
+    Saves sum to first row in the column
+*/
+template< typename T >
+__global__ void sum_columns(data_ptr< T > in, int x) {
+    const int col = blockIdx.x;
+    __shared__ warp_shared_type< T > sdata;
+
+    load_warp(in, sdata, [col, x](int tid) { return col + tid * x; });
+    sum_warp(in, sdata, col);
+}
+
+template< typename T, int shift, bool is_last >
+__device__ void load_to_shared(data_ptr< T > in, shared_type< T > sdata, int x, int block_col) {
+    const T gain = shift ? 6.0 : 1.0;
+    const int tid_x = threadIdx.x;
+    const int block_row = blockIdx.x * blockDim. x * x;
+
+    for (int j = 0; j < BLOCK_SIZE; ++j) {
+        if (is_last && (block_col + tid_x >= x)) {
+            break;
+        }
+        sdata[j][tid_x + shift] = in[block_row + block_col + x * j + tid_x] * gain;
+    }
     __syncthreads();
-    x = tiley + threadIdx.x; // transpose tiles
-    y = tilex + threadIdx.y; // transpose tiles
-    if (x >= ydim) return; // output matrix has y columns
-    int maxJ = min(32, xdim - y); // output matrix has x rows
-    for (int j = 0; j < maxJ; j += 8) {
-        int index = (y+j) * ydim + x;
-        odata[index] = tile[threadIdx.x][threadIdx.y + j];
+}
+
+template< typename T, int shift, bool is_last >
+__device__ void store_to_global(data_ptr< T > in, shared_type< T > sdata, int x, int block_col) {
```

Comment:
Again, why do I see this file? It is in the devel branch already. 
Are you sure there's no conflict?
Maybe consider rebase to the devel head?